### ISSUE TITLE
Assertion `replica->applier_sync_state == APPLIER_CONNECTED' failed.

### ISSUE THREAD

**Issue by rosik** [tarantool/tarantool#3510]
I've seen two errors today:
```
tarantool: /opt/tntsrc/src/box/replication.cc:234: void replica_on_applier_sync(replica*): Assertion `replica->applier_sync_state == APPLIER_CONNECTED' failed.
Aborted
```

```
tarantool: /opt/tntsrc/src/box/applier.cc:748: void applier_pause(applier*): Assertion `fiber() == applier->reader' failed.                                                           
Aborted
```

Tarantool version is 1.10.1-135-g193ef4150

It does not reproduce every time, and I can't give a minimal reproducible example, but
my workflow is to join several replicas (3rd or 4th) to vshard storage replicaset.

I also attact logs from all instances:
[logs.zip](https://github.com/tarantool/tarantool/files/2170887/logs.zip)

**Comment by locker** [comment]
The issue can be reproduced with the following script:
```
box.cfg{
    log_level = 4,
    listen = 44440 + arg[1],
    replication = {44441, 44442}
}
if box.info.id == 1 then
    box.schema.user.grant('guest', 'replication')
    box.space._cluster:delete(2)
end
os.exit(0)
```
Run it from two terminals:
- Terminal 1: `tarantool reproduce.lua 1`
- Terminal 2: `tarantool reproduce.lua 2`

One of the instances will crash:
```
2018-07-17 19:37:57.914 [19674] main/101/test.lua C> Tarantool 1.9.1-52-g1f187cacaabb
2018-07-17 19:37:57.914 [19674] main/101/test.lua C> log level 4
2018-07-17 19:37:57.926 [19674] main/105/applier/ xrow.c:792 E> ER_LOADING: Instance bootstrap hasn't finished yet
2018-07-17 19:37:58.928 [19674] main/105/applier/ coio.cc:104 !> SystemError connect, called on fd 12, aka [::1]:33652: Connection refused
tarantool: /home/vlad/src/tarantool-1.9/src/box/replication.cc:334: void replica_on_applier_disconnect(replica*): Assertion `0' failed.
Aborted (core dumped)
```

**Comment by sergepetrenko** [comment]
@rosik Hi
> ```
> tarantool: /opt/tntsrc/src/box/applier.cc:748: void applier_pause(applier*): Assertion `fiber() == applier->reader' failed.                                                           
> Aborted
> ```
Have you seen this assertion fail lately? What I tried is launching vshard example from the docs multiple times, but couldn't get such an error.
Maybe you can give me some of the more recent examples of this crash?

**Comment by locker** [comment]
This crash
```
tarantool: /opt/tntsrc/src/box/applier.cc:748: void applier_pause(applier*): Assertion `fiber() == applier->reader' failed.
```
was fixed in the scope of #3606.

**Comment by locker** [comment]
The other assertion
```
tarantool: /opt/tntsrc/src/box/replication.cc:234: void replica_on_applier_sync(replica*): Assertion `replica->applier_sync_state == APPLIER_CONNECTED' failed.
```
was caught here: https://github.com/tarantool/tarantool/issues/3610#issuecomment-410322819

So I guess we can close this issue as soon as the pending patch gets committed.

**Comment by sergepetrenko** [comment]
Tests on the latest commit hang due to a bug in test-run: tarantool/test-run#109

**Comment by locker** [comment]
Fixed by c939ca8aad365a19f8da3a835238f76b61cefec8
